File: bench/Reference/channel_throughput.java
import java.util.concurrent.*;

public class channel_throughput {
    static final Integer SENTINEL = Integer.MIN_VALUE;
    public static void main(String[] args) throws Exception {
        BlockingQueue<Integer> ch = new ArrayBlockingQueue<>(64);
        Thread producer = new Thread(() -> {
            try {
                for (int n = 1; n <= 10000; n++) ch.put(n);
                ch.put(SENTINEL);
            } catch (InterruptedException ignored) {}
        });
        producer.start();
        long sum = 0;
        Integer v;
        while (!(v = ch.take()).equals(SENTINEL)) sum += v;
        producer.join();
        System.out.println(sum);
    }
}

File: bench/Reference/channel_throughput.py
"""Bounded channel throughput benchmark — Python queue.Queue."""

import queue
import threading

CAP = 64
N = 10000


def producer(q):
    for i in range(1, N + 1):
        q.put(i)
    q.put(None)  # sentinel


def main():
    q = queue.Queue(CAP)
    t = threading.Thread(target=producer, args=(q,))
    t.start()
    s = 0
    while True:
        v = q.get()
        if v is None:
            break
        s += v
    t.join()
    print(s)


if __name__ == "__main__":
    main()
